lib/icinga/checkable-dependency.cpp

```diff
@@ -7,22 +7,181 @@
+ * @param refreshGlobalRegistry Whether to automatically refresh the global registry or not.
+ * @param modifiedGroup The dependency group that has been modified.
+ */
+void Checkable::AddDependency(const Dependency::Ptr& dependency, bool refreshGlobalRegistry, DependencyGroup::Ptr* modifiedGroup)
+{
+	std::lock_guard lock(m_DependencyMutex);
+	DependencyGroup::Ptr newGroup(new DependencyGroup(dependency->GetRedundancyGroup(), dependency));
+	if (auto it(m_DependencyGroups.find(newGroup)); it == m_DependencyGroups.end()) {
+		m_DependencyGroups.emplace(refreshGlobalRegistry ? DependencyGroup::Register(newGroup) : newGroup);
+	} else if (!refreshGlobalRegistry) {
+		// If we're not going to refresh the global registry, we just need to add the dependency to the existing group.
+		// Meaning, the dependency group itself isn't registered globally yet, so we don't need to re-register it.
+		(*it)->AddMember(dependency);
+
+		if (modifiedGroup) {
+			*modifiedGroup = *it;
+		}
+	} else {
+		if (auto existingGroup(*it); existingGroup->HasIdenticalMember(dependency)) {
+			// There's already an identical member in the group and this is likely an exact duplicate of it,
+			// so it won't change the identity of the group after registration, i.e. regardless whether we're
+			// supposed to refresh the global registry or not it's identity will remain the same.
+			existingGroup->AddMember(dependency);
+		} else {
+			// We're going to either replace the existing group with "newGroup" or merge the two groups together.
+			// Either way, it's identity will change, so we need to decouple it from the current Checkable.
+			m_DependencyGroups.erase(it);
+
+			if (modifiedGroup) {
+				// In both of the below cases, the identity of the existing group is going to probably change,
+				// so we'll need to clean up all the database entries/relations referencing its old identity.
+				*modifiedGroup = existingGroup;
+			}
+
+			if (auto members(existingGroup->GetMembers(this)); members.size() == existingGroup->GetMemberCount()) {
+				// We need to unregister the existing group from the global registry, as it's hash might change
+				// after adding the new dependency to it down below, and we want to re-register it afterwards.
+				// This way, we'll also be able to eliminate the possibility of having two identical groups in
+				// the registry that might occur due to the registration of the new dependency object below.
+				DependencyGroup::Unregister(existingGroup);
+				// The current Checkable is the only member of the group, so nothing to move around, just
+				// add the _duplicate_ dependency to the existing group. Duplicate in the sense that it's
+				// not identical to any of the existing members but similar enough to be part of the same
+				// group, i.e. same parent, maybe different period, state filter, etc.
+				existingGroup->AddMember(dependency);
+				m_DependencyGroups.emplace(DependencyGroup::Register(existingGroup));
```

Comment:
There's a possible race condition here as the `if` check here happens holding just the per-checkable `m_DependencyMutex` and another call could get the group from the registry and modify it in between.